internal/query/relations.go

```diff
@@ -803,9 +1048,35 @@ func (e *engine) DeleteRole(ctx context.Context, roleResource types.Resource) er
 		}
 	}
 
+	dbTx, err := e.db.DeleteRoleTransaction(ctx, roleResource.ID)
+	if err != nil {
+		// If the role doesn't exist, simply ignore.
+		if !errors.Is(err, database.ErrNoRoleFound) {
+			return err
+		}
+	} else {
+		// Setup rollback in case an error occurs before we commit.
+		defer dbTx.Rollback() //nolint:errcheck // error is logged in function
+	}
+
 	for _, filter := range filters {
 		if err = e.deleteRelationships(ctx, filter); err != nil {
-			return fmt.Errorf("failed to delete role action %s: %w", filter.OptionalResourceId, err)
+			err = fmt.Errorf("failed to delete role action %s: %w", filter.OptionalResourceId, err)
+
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+
+			return err
+		}
+	}
```

Comment:
What happens if I attempt to update a role with new actions while it's being deleted? Reading this code, it looks possible that I wind up with relationships in SpiceDB but no corresponding CRDB entries. This might also, ironically, require a `SELECT ... FOR UPDATE`.

Reply:
this is now handled properly as we always have a transaction open before on the database. if two changes came in at the same time, one would have to wait until the database released the row.